src/backend/distributed/utils/resource_lock.c

```diff
@@ -1060,3 +1061,143 @@ CitusLockTableAclCheck(Oid relationId, LOCKMODE lockmode, Oid userId)
 
 	return aclResult;
 }
+
+
+/*
+ * EnsureCanAcquireLock checks if currect user has the permissions
+ * to acquire a lock on the table and throws an error if the user does
+ * not have the permissions
+ */
+static void
+EnsureCanAcquireLock(Oid relationId, LOCKMODE lockMode)
+{
+	AclResult aclResult = CitusLockTableAclCheck(relationId, lockMode,
+												 GetUserId());
+	if (aclResult != ACLCHECK_OK)
+	{
+		aclcheck_error(aclResult,
+					   get_relkind_objtype(get_rel_relkind(relationId)),
+					   get_rel_name(relationId));
+	}
+}
+
+
+/*
+ * AcquireDistributedLockOnRelations acquire a distributed lock on worker nodes
+ * for given list of relations ids. Worker node list is sorted so that the lock
+ * is acquired in the same order regardless of which node it was run on. Notice
+ * that no lock is acquired on coordinator node if the coordinator is not added
+ * to the metadata.
+ */
+void
+AcquireDistributedLockOnRelations(List *relationIdList, LOCKMODE lockMode)
+{
+	if (!EnableMetadataSync)
+	{
+		return;
+	}
+
+	const char *lockModeText = LockModeToLockModeText(lockMode);
+
+	UseCoordinatedTransaction();
+
+	StringInfo lockRelationsCommand = makeStringInfo();
+	appendStringInfo(lockRelationsCommand, "%s;\n", DISABLE_DDL_PROPAGATION);
+
+	List *relationIdsToLock = NIL;
+	bool partialLockStmt = false;
+
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, relationIdList)
+	{
+		EnsureCanAcquireLock(relationId, lockMode);
+
+		CitusTableCacheEntry *tableCacheEntry = GetCitusTableCacheEntry(relationId);
+
+		/* skip citus tables which are not seen by workers */
+		if (tableCacheEntry->partitionMethod == DISTRIBUTE_BY_RANGE ||
+			tableCacheEntry->partitionMethod == DISTRIBUTE_BY_APPEND)
+		{
+			continue;
+		}
+
+		char *qualifiedRelationName = generate_qualified_relation_name(relationId);
+
+		/*
+		 * As of pg14 we cannot use LOCK to lock a FOREIGN TABLE
+		 * so we have to use the lock_relation_if_exist udf
+		 */
+		if (get_rel_relkind(relationId) == RELKIND_FOREIGN_TABLE)
+		{
+			/* finish the partial lock statement */
+			if (partialLockStmt)
+			{
+				appendStringInfo(lockRelationsCommand, " IN %s MODE;\n", lockModeText);
+				partialLockStmt = false;
+			}
+
+			/* use lock_relation_if_exits to lock foreign table */
+			appendStringInfo(lockRelationsCommand, LOCK_RELATION_IF_EXISTS,
+							 quote_literal_cstr(qualifiedRelationName),
+							 quote_literal_cstr(lockModeText));
+			appendStringInfoChar(lockRelationsCommand, '\n');
+		}
+		else if (partialLockStmt)
```

Comment:
minor: maybe rename `partialLockStmt` to `startedLockCommand` or such? And, document it with some comments like:
```
/* 
 * In this loop, when there are foreign tables, we need to switch from LOCK * statement to lock_relation_if_exists() as foreign tables do not support the * former. 
 */
bool startedLockCommand = false;